src/web100srv.c

```diff
@@ -2614,12 +2446,57 @@ int main(int argc, char** argv) {
                   case 'S':
                     SysLogFacility = optarg;
                     break;
+                  case 314:
+                    options.tls = 1;
+                    break;
+                  case 315:
+                    private_key_file = optarg;
+                    break;
+                  case 316:
+                    certificate_file = optarg;
+                    break;
                   case '?':
                     short_usage(argv[0], "");
                     break;
                 }
               }
 
+  if (options.tls) {
+    if (private_key_file == NULL || certificate_file == NULL) {
+      short_usage(argv[0], "TLS requires a private key and a certificate");
+    }
+
+    SSL_library_init();
+    OpenSSL_add_all_algorithms();
+    SSL_load_error_strings();
+    ssl_context = SSL_CTX_new(SSLv23_server_method());
+    if (ssl_context == NULL) {
+      ssl_err = ERR_get_error();
+      ERR_error_string_n(ssl_err, ssl_error, sizeof(ssl_error));
+      printf("SSL_CTX_new: %s\n", ssl_error);
+      short_usage(argv[0], "SSL/TLS context could not be created.");
+    }
+    SSL_CTX_set_mode(ssl_context, SSL_MODE_AUTO_RETRY);
+    // SSL private key file initialization goes here
+    if (SSL_CTX_use_certificate_file(ssl_context, certificate_file,
+                                     SSL_FILETYPE_PEM) != 1) {
+      ssl_err = ERR_get_error();
+      ERR_error_string_n(ssl_err, ssl_error, sizeof(ssl_error));
+      printf("SSL_CTX_use_certificate_file: %s\n", ssl_error);
+      short_usage(argv[0], "SSL/TLS certificate file could not be loaded.");
+    }
+    if (SSL_CTX_use_PrivateKey_file(ssl_context, private_key_file,
+                                    SSL_FILETYPE_PEM) != 1) {
+      ssl_err = ERR_get_error();
+      ERR_error_string_n(ssl_err, ssl_error, sizeof(ssl_error));
+      printf("SSL_CTX_use_PrivateKey_file: %s\n", ssl_error);
+      short_usage(argv[0], "SSL/TLS private key file could not be loaded.");
+    }
+    if (!SSL_CTX_check_private_key(ssl_context)) {
+      short_usage(argv[0], "Private key does go with certificate.");
+    }
```

Comment:
Can we refactor 2465-2497 into its own function instead of further expanding `main`?